apps/designer/app/designer/features/props-panel/control.tsx

```diff
@@ -117,109 +160,186 @@ const BooleanControl = ({
 };
 
-type RangeControlProps = BaseControlProps<number> & {
-  type: "range";
-  min: number;
-  max: number;
-  step: number;
+const assertUnreachable = (_arg: never) => {
+  throw new Error("Didn't expect to get here");
 };
 
-type OptionsControlProps = BaseControlProps & {
-  type:
-    | "radio"
-    | "inline-radio"
-    | "check"
-    | "inline-check"
-    | "select"
-    | "multi-select";
-  options: Array<string>;
+type ControlProps = {
+  component: Instance["component"];
+  userProp: UserProp;
+  onChangePropValue: (value: string | boolean) => void;
+  onChangePropAsset: (asset: Asset) => void;
 };
 
-export type ControlProps =
-  | PrimitiveControlProps
-  | OptionsControlProps
-  | FileControlProps
-  | ColorControlProps
-  | RangeControlProps;
+const optionsControlTypes = [
+  ...radioControlTypes,
+  ...checkboxControlTypes,
+  ...selectControlTypes,
+];
 
 // eslint-disable-next-line func-style
-export function Control(props: ControlProps) {
-  switch (props.type) {
-    case "array":
-      return <TextControl {...props} />;
-    case "boolean":
-      return <BooleanControl {...props} />;
-    case "color":
-      return <TextControl {...props} type="color" />;
-    case "date":
-      return <TextControl {...props} type="date" />;
-    case "file":
-      return <TextControl {...props} type="file" />;
-    case "number":
-      return <TextControl {...props} type="number" />;
-    case "range":
-      return <RangeControl {...props} />;
-    case "object":
-      return <TextControl {...props} />;
-    case "radio":
-      return <RadioControl {...props} />;
-    case "inline-radio":
-      return <RadioControl {...props} />;
-    case "check":
-      return <CheckboxControl {...props} />;
-    case "inline-check":
-      return <CheckboxControl {...props} />;
-    case "select":
-      return <SelectControl {...props} />;
-    case "multi-select":
-      return <CheckboxControl {...props} />;
-    case "text":
-      return <TextControl {...props} />;
-    default: {
-      const _exhaustiveCheck: never = props;
-      return <NotImplemented />;
+export function Control({
+  component,
+  userProp,
+  onChangePropValue,
+  onChangePropAsset,
+}: ControlProps) {
+  const meta = componentsMeta[component];
+  const argType = meta.argTypes?.[userProp.prop as keyof typeof meta.argTypes];
+
+  const defaultValue = argType?.control.defaultValue;
+  const type = argType?.control.type;
+
+  if (type == null) {
+    warnOnce(
+      true,
+      `No control type for prop: ${userProp.prop} component: ${component} found`
+    );
+    return <NotImplemented />;
+  }
+
+  if (typeof type !== "string") {
+    warnOnce(
+      true,
+      `Control type for prop: ${
+        userProp.prop
+      } component: ${component} is not a string but ${typeof type}`
+    );
+
+    return <NotImplemented />;
+  }
+
+  if (isFeatureEnabled("assets")) {
+    if (component === "Image" && userProp.prop === "src") {
```

Comment:
Why we check `component` and `userProp.prop` directly rather than going via `const type = argType?.control.type` like for all other control types? Is this temporary?

Reply:
Because we want this only for selected component `Image`, for known prop name 'src' only.

`type` there is coming from storybook and is `string` or like, there are no type `asset` or like on our components.